src/effects/VST3/internal/ComponentHandler.cpp

```diff
@@ -0,0 +1,123 @@
+/**********************************************************************
+
+  Audacity: A Digital Audio Editor
+
+  @file ComponentHandler.cpp
+
+  @author Vitaly Sverchinsky
+
+  @brief Part of Audacity VST3 module
+
+**********************************************************************/
+
+#include "ComponentHandler.h"
+
+#include <pluginterfaces/vst/ivstparameterchanges.h>
+
+#include "ParameterChanges.h"
+
+internal::ComponentHandler::ComponentHandler()
+{
+   //We use two instances of ParameterChanges to ensure
+   //that read/write is lock and wait free.
+   //Once parameter value change is requested, they are
+   //written either to the pending outgoing parameters,
+   //if they are not yet requested, or to any of those
+   //two instances whichever of them is available.
+   //When done writing, parameter changes become pending.
+
+   using namespace Steinberg;
+
+   FUNKNOWN_CTOR
+   auto first = owned(safenew internal::ParameterChanges);
+   auto second = owned(safenew internal::ParameterChanges);
+   if(first && second)
+   {
+      mFirst = first.take();
+      mSecond = second.take();
+   }
+}
+
+internal::ComponentHandler::~ComponentHandler()
+{
+   Steinberg::Vst::IParameterChanges* ptr { nullptr };
+   if(ptr = mFirst.load())
+      ptr->release();
+   if(ptr = mSecond.load())
+      ptr->release();
+   if(ptr = mPendingChanges.load())
+      ptr->release();
+}
+
+Steinberg::tresult internal::ComponentHandler::beginEdit(Steinberg::Vst::ParamID)
+{
+   return Steinberg::kResultOk;
+}
+
+Steinberg::tresult internal::ComponentHandler::performEdit(Steinberg::Vst::ParamID id,
+   Steinberg::Vst::ParamValue valueNormalized)
+{
+   using namespace Steinberg;
+
+   //Try grab pending changes first (processing thread will not
+   //see them until they are applied)
+   auto lock = mPendingChanges.exchange(nullptr);
+   if(lock == nullptr)
+   {
+      //There is no pending changes, then try grab
+      //any of the containers. At least one of them always should be
+      //available, but another one may be grabbed by the processing thread
+      lock = mFirst.exchange(nullptr);
+      if(lock == nullptr)
+         lock = mSecond.exchange(nullptr);
+
+      //Grabbed object may contain some old data, clean it.
+      //We surely don't want to delay the processing thread with cleanup
+      //routines
+      static_cast<ParameterChanges*>(lock)->clear();
+   }
```

Comment:
I think I missed some interesting discussions during my week off.

What's the proof of correctness of this algorithm?

1. Tthere are two `ParameterChanges` objects, and at all times each is held in exactly one of five places (editor, processor, first, second, or pending) and each place holds at most one.
2. Only the editor thread ever makes pending non-null.
3. Only the editor thread ever makes second null.

Every exchange satisfies those conditions; from that you can deduce that editor never fails to lock one of the objects.

Do I reason right?  Is this reasoning worth commenting?

Reply:
See also the class MessageBuffer in PlaybackSchedule.h where I solved the same problem in a lock-free, but not wait-free algorithm, yet very probably looping only very rarely, and using only two slots.

Which of these solutions is better?

I think there is a utility to be extract from it into one of the lower libraries.

Which are the threads that communicate?

In the cases of my MessageBuffer, it's the main thread sending messages to a worker thread -- the one which must maintain high throughput but can tolerate variation in latency, not the one that executes as the callback for the PortAudio library.  Changes of looping play bounds are the sort of changes done there.

In your case, I presume this is also about adjustment of sliders detected by the main thread.  It is the low-latency thread that is receiving the messages.  However, I believe that is a wrong implementation (an old error and nothing wrong with this PR).  See issue #2082.  Things should be restructured so the receiver thread for your messages is the same as those using MessageBuffer.

Reply:
> I think I missed some interesting discussions during my week off.
> 
> What's the proof of correctness of this algorithm?
> 
> 1. Tthere are two `ParameterChanges` objects, and at all times each is held in exactly one of five places (editor, processor, first, second, or pending) and each place holds at most one.
> 2. Only the editor thread ever makes pending non-null.
> 3. Only the editor thread ever makes second null.
> 
> Every exchange satisfies those conditions; from that you can deduce that editor never fails to lock one of the objects.
> 
> Do I reason right? Is this reasoning worth commenting?


1 - yes, 2 - yes, 3 - yes, and first one as well.
There are only two threads (at most): processing and UI/Editor thread. There are only two buffers that are used to store changes, any of them is either: 
1) stored in `mFirst` or `mSecond` and considered to be empty, UI thread may move one of them to `mPendingChanges`
2) stored in `mPendingChanges`, only UI thread is allowed to store on of the buffers in that variable.
3) stored in unique pointer returned by `getPendingChanges`, and isn't accessible from UI thread until buffer is returned to `mFirst` or `mSecond` by processing thread

UI thread will try to lock the buffer in `mPendingChanges` first, if it locked one in `mFirst` or `mSecond` instead it may happen so that `mPendingChanges` isn't yet read by processing thread and then the implementation may become a bit more complicated.
At most one of the buffers is hold by processing thread, at least one of the buffer is available from UI thread (`mPendingChanges`, `mFirst` or `mSecond`)